### PR TITLE
Texture region splitting

### PR DESCRIPTION
**Pull request by sabretooth**
Implementing a potential fix for #795 by generating cropped views of textures on-demand.

### CODE REVIEW

File: src/xenia/gpu/vulkan/vulkan_command_processor.cc [hunk 1]
```diff
@@ -854,7 +856,7 @@ bool VulkanCommandProcessor::PopulateSamplers(VkCommandBuffer command_buffer,
 
   std::vector<xe::gpu::Shader::TextureBinding> dummy_bindings;
   auto descriptor_set = texture_cache_->PrepareTextureSet(
-      setup_buffer, current_batch_fence_, vertex_shader->texture_bindings(),
+      command_buffer, current_batch_fence_, vertex_shader->texture_bindings(),
```
Comment (DrChat) [inline src/xenia/gpu/vulkan/vulkan_command_processor.cc]: Q: Why did you change this to command_buffer? AFAIK uploads cannot be completed inside of a render pass.

File: src/xenia/gpu/vulkan/render_cache.cc [hunk 2]
```diff
@@ -713,6 +716,28 @@ const RenderState* RenderCache::BeginRenderPass(VkCommandBuffer command_buffer,
   vkCmdBeginRenderPass(command_buffer, &render_pass_begin_info,
                        VK_SUBPASS_CONTENTS_INLINE);
 
+  // XXX HACK: clear the color target if the previous render pass was a
+  // depth-stencil pass targeting the same EDRAM tile
+  // (some games clear color targets with a depth pass, since it's slightly
+  // faster)
+  if (previous_render_pass && previous_render_pass->config.depth_stencil.used &&
```
Comment (DrChat) [inline src/xenia/gpu/vulkan/render_cache.cc]: This may need to be axed for now. It's causing a regression in banjo-kazooie for me as of now.

**Comment by DrChat** [comment]
Good work! Just need a couple of things addressed for now.

File: src/xenia/gpu/vulkan/vulkan_command_processor.cc (hunk 1 above)
Comment (sabretooth) [reply src/xenia/gpu/vulkan/vulkan_command_processor.cc]: The setup_buffer is built through the entire frame and executed in its entirety before the command_buffer. The blits that we add here for carving up the resolve textures depend on the render passes completed in command_buffer; putting them in setup_buffer would have the blits execute before the render targets they read from are written.

File: src/xenia/gpu/vulkan/render_cache.cc (hunk 2 above)
Comment (sabretooth) [reply src/xenia/gpu/vulkan/render_cache.cc]: Good to know, I'll look at that game and see if I can fix the regression.

File: src/xenia/gpu/vulkan/texture_cache.cc [hunk 3]
```diff
@@ -325,17 +350,21 @@ bool TextureCache::FreeTexture(Texture* texture) {
     vkDestroyFramebuffer(*device_, texture->framebuffer, nullptr);
   }
 
-  for (auto it = texture->views.begin(); it != texture->views.end();) {
-    vkDestroyImageView(*device_, (*it)->view, nullptr);
-    it = texture->views.erase(it);
+  for (auto region_it = texture->regions.begin();
```
Comment (DrChat) [inline src/xenia/gpu/vulkan/texture_cache.cc]: You don't appear to be freeing the base_region here

File: src/xenia/gpu/vulkan/vulkan_command_processor.cc (hunk 1 above)
Comment (DrChat) [reply src/xenia/gpu/vulkan/vulkan_command_processor.cc]: Okay, that's fair but you'll need to end the render pass when doing an upload/blit (as per vulkan spec).

I'd say pass both the setup buffer and command buffer, do uploads from memory in the setup buffer and uploads from other textures in the command buffer.

File: src/xenia/gpu/vulkan/texture_cache.cc (hunk 3 above)
Comment (DrChat) [reply src/xenia/gpu/vulkan/texture_cache.cc]: Also - there's a regression somewhere causing a texture to be freed twice in this branch, but I haven't tracked it down quite yet.

File: src/xenia/gpu/vulkan/texture_cache.cc [hunk 4]
```diff
@@ -404,13 +433,23 @@ TextureCache::Texture* TextureCache::DemandResolveTexture(
       },
       this, texture);
 
+  // Invalidate any textures that share this memory location
+  for (auto it = textures_.begin(); it != textures_.end(); ++it) {
```
Comment (DrChat) [inline src/xenia/gpu/vulkan/texture_cache.cc]: This block of code is causing a texture to be freed twice - it appears that an access watch fires and then this block of code runs over the texture as well.

**Comment by DrChat** [comment]
Update: Keep an eye on [the texsplit branch](https://github.com/benvanik/xenia/tree/texsplit) - I'm making changes there.

**Comment by DrChat** [comment]
Continued in new PR.
